Added logging for the Reducer's non-member functions.

averywang21 opened the pull request:
Context: https://github.com/pytorch/pytorch/issues/59338

Summary:
Added an optional logging parameter for non-member functions `compute_bucket_assignment_by_size` and `verify_replica0_across_processes`. If a logger is provided then `TORCH_CHECK` assertions are replaced with a wrapper that logs the error to the DDP reducer's logger before calling `TORCH_CHECK`. If a logger is not provided `TORCH_CHECK` is still called.

Modified python-side calls to `_compute_bucket_assignment_by_size` and `_verify_model_across_ranks` to include a logger whenever possible. A notable exception is when these non-member functions are called in DDP's constructor - we cannot pass in a logger as they may have not been initialized yet.

We also added 4 new tests: `test_compute_bucket_assignment_by_size_sparse_error_{with, without}_logger` which tests the `_compute_bucket_assignment_by_size` function to ensure that sparse tensors are rejected and the errors are logged.  `test_verify_model_across_rank_{with, without}_logger` calls `_verify_model_across_ranks` to ensure that ill-formed models (different ranks have different number of parameters compared to rank 0) are rejected and the errors are logged. The test `test_ddp_model_diff_across_ranks` remains unchanged - while it does construct a ill-formed DDP instance which triggers the error in `_verify_model_across_ranks`, we cannot check the logger because this error appears in the constructor.

Lastly, did some cleanup of the `test_ddp_model_diff_across_ranks` function to make the logic of choosing which context manager and error message to use more clean.

Test Plan:
**Build commands**
`buck build mode/dev-nosan //caffe2/test/distributed:distributed_nccl_spawn --keep-going`

`buck build mode/dev-nosan //caffe2/test/distributed:distributed_gloo_spawn --keep-going`

**Test commands**
Test for `_compute_bucket_assignment_by_size` (Python)/ `compute_bucket_assignment_by_size` (C++)
`BACKEND={nccl, gloo} WORLD_SIZE=2 ../buck-out/dev/gen/caffe2/test/distributed/distributed_{nccl, gloo}_spawn#binary.par -r test_compute_bucket_assignment_by_size_sparse_error_{with, without}_logger`

Test for `_verify_model_across_ranks` (Python)/`verify_replicas0_across_process` (C++)
`BACKEND={nccl, gloo} WORLD_SIZE=2 ../buck-out/dev/gen/caffe2/test/distributed/distributed_{nccl, gloo}_spawn#binary.par -r test_verify_model_across_ranks_{with, without}_logger`

Test that constructs an ill-formed DDP instance. Only did cleanup of this function.
`BACKEND={nccl, gloo} WORLD_SIZE=2 ../buck-out/dev/gen/caffe2/test/distributed/distributed_{nccl, gloo}_spawn#binary.par -r test_ddp_model_diff_across_ranks`

Differential Revision: D30924790



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @cbalioglu @gcramer23

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1544,18 +1544,40 @@ Example::
+            const std::vector<bool>& expect_sparse_gradient,
+            const std::vector<int64_t>& tensor_indices,
+            const c10::optional<std::weak_ptr<::c10d::Logger>>& logger) {
+             if (logger.has_value()) {
```
Great job figuring out the pybind stuff so quickly! 

You've modified `c10d::compute_bucket_by_size` to already take an optional logger, so I don't think we need the if/else branch here. If `logger` doesn't have the value set it will be appropriately handled in the reducer so just pass `logger` regardless here?

rohan-varma replied:
Nevermind, forgot that we need this to be weak_ptr. The branching based on `has_value` looks fine to me. 

Can we make this pybind similar to `set_logger` pybind, but it will have c10::optional? i.e. pass in 

`const c10::optional<std::shared_ptr<::c10d::Logger>> logger`.

rohan-varma commented:
note to reviewers - Longer term, we'll need to refactor the reducer so that we can call things like `_verify_model_across_ranks` with a valid logger, which will increase the usefulness of error logging. However it is out of scope for this PR and this PR can serve as the basic building blogs of adjusting the reducer logic appropriately if there is a logger.

SciPioneer commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1990,7 +1992,13 @@ compute_bucket_assignment_by_size(
     // TODO: This is not a reducer method so it does not have access to logger,
     // pass in logger directly here.
-    TORCH_CHECK(!tensor.is_sparse(), "No support for sparse tensors.");
+
```
nit: don't need the empty lines

SciPioneer commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -2079,7 +2087,8 @@ compute_bucket_assignment_by_size(
 void verify_replica0_across_processes(
     c10::intrusive_ptr<c10d::ProcessGroup> process_group,
-    std::vector<std::vector<at::Tensor>> model_replicas) {
+    std::vector<std::vector<at::Tensor>> model_replicas,
```
nit: pass by const reference? Since each tensor has metadata that won't be shared automatically like storage, still better to use reference here.

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6874,38 +6874,152 @@ def test_ddp_model_diff_across_ranks(self):
             torch.cuda.set_device(self.rank)
             # Creates network with different sized embedding table on different
```
Update the comment accordingly?

SciPioneer replied:
Update the comment accordingly?

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6874,38 +6874,152 @@ def test_ddp_model_diff_across_ranks(self):
             )
+            torch.cuda.set_device(self.rank)
+            # Creates network with different sized embedding table on different
+            # ranks. This should throw an error during DDP init.
```
Ditto

SciPioneer commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -602,11 +602,13 @@ compute_bucket_assignment_by_size(
 TORCH_API void verify_replica0_across_processes(
     c10::intrusive_ptr<c10d::ProcessGroup> process_group,
-    std::vector<std::vector<at::Tensor>> model_replicas);
+    std::vector<std::vector<at::Tensor>> model_replicas,
```
Ditto

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6874,38 +6874,152 @@ def test_ddp_model_diff_across_ranks(self):
+            # rather, it will be taken down by NCCL_ASYNC_ERROR_HANDLING. When
+            # running with Gloo or with debug mode wrapper, we expect the error
+            # to be caught inline.
+            is_detail_dbg_mode = (
```
Probably can be wrapped as a helper function.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6874,38 +6874,152 @@ def test_ddp_model_diff_across_ranks(self):
                     device_ids=[self.rank],
                     process_group=group_to_use,
                 )
+
```
remove empty line

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6874,38 +6874,152 @@ def test_ddp_model_diff_across_ranks(self):
+                    expected_err = "Caught collective operation timeout"
+                    ctx = self.assertRaisesRegex(RuntimeError, expected_err)
+                else:
+                    expected_err = ""
```
Don't need to set this in this branch right?

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6874,38 +6874,152 @@ def test_ddp_model_diff_across_ranks(self):
+                device_ids=[self.rank],
+                process_group=group_to_use,
+            )
+            net.module.lin = nn.Linear(100 if self.rank == 0 else 10, 1)
```
Comment why we're doing this (ie. to make the model different and test error)

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6874,38 +6874,152 @@ def test_ddp_model_diff_across_ranks(self):
-                rank_0_ctx
-                if self.rank == 0
-                else self.assertRaisesRegex(RuntimeError, "appears not to match")
+            net.module.lin = nn.Linear(100 if self.rank == 0 else 10, 1)
```
same - comment why

rohan-varma reviewed (approved):
Great work! Please wait for all internal and OSS tests to pass before landing .

rohan-varma commented:
CI failures are unrelated.

averywang21 committed:
Added logging for the Reducer's non-member functions. (#65023)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/65023

Added an optional logging parameter for non-member functions `compute_bucket_assignment_by_size` and `verify_replica0_across_processes`. If a logger is provided then `TORCH_CHECK` assertions are replaced with a wrapper that logs the error to the DDP reducer's logger before calling `TORCH_CHECK`. If a logger is not provided `TORCH_CHECK` is still called.

Modified python-side calls to `_compute_bucket_assignment_by_size` and `_verify_model_across_ranks` to include a logger whenever possible. A notable exception is when these non-member functions are called in DDP's constructor - we cannot pass in a logger as they may have not been initialized yet.

We also added 4 new tests: `test_compute_bucket_assignment_by_size_sparse_error_{with, without}_logger` which tests the `_compute_bucket_assignment_by_size` function to ensure that sparse tensors are rejected and the errors are logged.  `test_verify_model_across_rank_{with, without}_logger` calls `_verify_model_across_ranks` to ensure that ill-formed models (different ranks have different number of parameters compared to rank 0) are rejected and the errors are logged. The test `test_ddp_model_diff_across_ranks` remains unchanged - while it does construct a ill-formed DDP instance which triggers the error in `_verify_model_across_ranks`, we cannot check the logger because this error appears in the constructor.

Lastly, did some cleanup of the `test_ddp_model_diff_across_ranks` function to make the logic of choosing which context manager and error message to use more clean.

Test Plan:
**Build commands**
`buck build mode/dev-nosan //caffe2/test/distributed:distributed_nccl_spawn --keep-going`

`buck build mode/dev-nosan //caffe2/test/distributed:distributed_gloo_spawn --keep-going`

**Test commands**
Test for `_compute_bucket_assignment_by_size` (Python)/ `compute_bucket_assignment_by_size` (C++)
`BACKEND={nccl, gloo} WORLD_SIZE=2 ../buck-out/dev/gen/caffe2/test/distributed/distributed_{nccl, gloo}_spawn#binary.par -r test_compute_bucket_assignment_by_size_sparse_error_{with, without}_logger`

Test for `_verify_model_across_ranks` (Python)/`verify_replicas0_across_process` (C++)
`BACKEND={nccl, gloo} WORLD_SIZE=2 ../buck-out/dev/gen/caffe2/test/distributed/distributed_{nccl, gloo}_spawn#binary.par -r test_verify_model_across_ranks_{with, without}_logger`

Test that constructs an ill-formed DDP instance. Only did cleanup of this function.
`BACKEND={nccl, gloo} WORLD_SIZE=2 ../buck-out/dev/gen/caffe2/test/distributed/distributed_{nccl, gloo}_spawn#binary.par -r test_ddp_model_diff_across_ranks`

Reviewed By: rohan-varma

Differential Revision: D30924790

fbshipit-source-id: b65a103d6121a4211549fc3922fe00510194c2b8
